### ISSUE TITLE
Unnecessary reflow due to setState in componentWillReceiveProps

### ISSUE THREAD

**Issue by univerio** [facebook/react#1940]
See [mailing list post](https://groups.google.com/d/msg/reactjs/pKxD3JY_Y84/zc59P2T5IP4J), reproduced below for convenience:

The following reduced test case ([jsfiddle](http://jsfiddle.net/Tn83a/)) has an unnecessary reflow:

``` javascript
var Input = React.createClass({
    getInitialState: function () {
        return {
            text: this.props.text + "..."
        };
    },
    componentWillReceiveProps: function (nextProps) {
        this.setState({
            text: nextProps.text + "..."
        });
    },
    render: function () {
        return React.DOM.div(null, this.state.text);
    }
});

var App = React.createClass({
    getInitialState: function () {
        return {
            value: "foo"
        };
    },
    render: function () {
        var that = this;
        return React.DOM.div(null,
            React.DOM.input({
                value: this.state.value,
                onChange: function (e) {
                    that.setState({
                        value: e.target.value
                    });
                }
            }),
            Input({
                text: this.state.value
            }));
    }
});

React.renderComponent(App(null), document.body);
```

And a screenshot from the Timeline tab in Chrome Dev Tools showing the reflow:

![reflow](https://cloud.githubusercontent.com/assets/1037172/3712415/33574b9e-1513-11e4-9dca-cecc3233353a.png)

An equivalent implementation ([jsfiddle](http://jsfiddle.net/EG7hL/)) not using `setState` in `componentWillReceiveProps` does not have the same reflow:

``` javascript
var Input = React.createClass({
    render: function () {
        return React.DOM.div(null, this.props.text + "...");
    }
});

var App = React.createClass({
    getInitialState: function () {
        return {
            value: "foo"
        };
    },
    render: function () {
        var that = this;
        return React.DOM.div(null,
            React.DOM.input({
                value: this.state.value,
                onChange: function (e) {
                    that.setState({
                        value: e.target.value
                    });
                }
            }),
            Input({
                text: this.state.value
            }));
    }
});

React.renderComponent(App(null), document.body);
```

**Comment by theREALdebater** [comment]
In a React v16 component, I defined componentWillReceiveProps inside which I always called setState (on an object with several members, including other objects). This seems to have caused a runaway loop of re-renders, because when I added code to make the call to setState conditional the loop was no longer observed. I thought calling setState in componentWillReceiveProps (and also componentWillMount) was not meant to trigger a re-render. Am I wrong? Has this changed in v16?
